database/src/main/java/com/firebase/ui/database/paging/FirebaseDataSource.java

```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Override
+    public void loadInitial(@NonNull LoadInitialParams<String> params, @NonNull final LoadInitialCallback<String, T> callback) {
+        // Set initial loading state
+        mLoadingState.postValue(LoadingState.LOADING_INITIAL);
+
+        Query mInitQuery = mQuery.limitToFirst(params.requestedLoadSize);
+        mInitQuery.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                if (dataSnapshot.exists()) {
+                    ArrayList<T> mDataList = new ArrayList<T>();
+                    for (DataSnapshot ds : dataSnapshot.getChildren()) {
+                        T data = ds.getValue(mClass);
+                        String key = ds.getKey();
+                        mKeyList.add(key);
+                        mDataList.add(data);
+                    }
+
+                    //Initial Load Success
+                    mLoadingState.postValue(LoadingState.LOADED);
+                    callback.onResult(mDataList, mKeyList.get(mKeyList.size() - 1), mKeyList.get(mKeyList.size() - 1));
+                }
+                else {
+                    setWrongDataPathError();
+                }
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError databaseError) {
+                //Error Occured
+                mError.postValue(databaseError);
+                mLoadingState.postValue(LoadingState.ERROR);
+            }
+        });
+    }
+
+    @Override
+    public void loadBefore(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, T> callback) {
+        // Ignored for now, since we only ever append to the initial load.
+    }
+
+    @Override
+    public void loadAfter(@NonNull LoadParams<String> params, @NonNull final LoadCallback<String, T> callback) {
+        // Set loading state
+        mLoadingState.postValue(LoadingState.LOADING_MORE);
+
+        //Load params.requestedLoadSize+1 because, first data item is getting ignored.
+        Query mNewQuery = mQuery.startAt(null, params.key).limitToFirst(params.requestedLoadSize + 1);
+        mNewQuery.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                if (dataSnapshot.exists()) {
+                    ArrayList<T> mList = new ArrayList<T>();
+                    boolean isFirstItem = true;
+
+                    for (DataSnapshot ds : dataSnapshot.getChildren()) {
+                        T data = ds.getValue(mClass);
+                        String key = ds.getKey();
+
+                    /*
+                      Check for first Item.
+                      Because in Firebase Database there is no query for startAfter(key).
+                      So we're ignoring first data item
+                    */
+                        if (!isFirstItem) {
```

Comment:
This loop seems over complicated.  Rather than checking `firstItem` in the loop over and over why not just iterate from child `1` to the end, skipping the first one?